api/layer/multipart_upload.go

```diff
@@ -0,0 +1,459 @@
+
+	for _, ov := range uploads {
+		if ov.IsDir {
+			result.Prefixes = append(result.Prefixes, ov.Key)
+		} else {
+			result.Uploads = append(result.Uploads, ov)
+		}
+	}
+
+	return &result, nil
+}
+
+func (n *layer) AbortMultipartUpload(ctx context.Context, p *UploadInfoParams) error {
+	objects, err := n.getUploadParts(ctx, p.UploadID, p.Key, p.Bkt)
+	if err != nil {
+		return err
+	}
+
+	for _, info := range objects {
+		err := n.objectDelete(ctx, info.CID, info.ID)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (n *layer) ListParts(ctx context.Context, p *ListPartsParams) (*ListPartsInfo, error) {
+	var res ListPartsInfo
+	uploads, err := n.getUploadParts(ctx, p.Info.UploadID, p.Info.Key, p.Info.Bkt)
+	if err != nil {
+		return nil, err
+	}
+	res.Owner = uploads[0].Owner
+
+	uploads = uploads[p.PartNumberMarker+1:]
+
+	if len(uploads) > p.MaxParts {
+		res.IsTruncated = true
+		res.NextPartNumberMarker = p.MaxParts + p.PartNumberMarker
+		uploads = uploads[:p.MaxParts+1]
+	}
+
+	res.Parts = uploads
+
+	return &res, nil
+}
+
+func (n *layer) GetUploadInitInfo(ctx context.Context, p *UploadInfoParams) (*data.ObjectInfo, error) {
+	filters := []*filter{
+		{attr: UploadIDAttributeName, val: p.UploadID},
+		{attr: UploadPartNumberAttributeName, val: "0"},
+		{attr: UploadKeyAttributeName, val: p.Key},
+	}
+	f := &findParams{
+		filters: filters,
+		cid:     p.Bkt.CID,
+	}
+	ids, err := n.objectSearch(ctx, f)
+	if err != nil {
+		if errors.IsS3Error(err, errors.ErrNoSuchKey) {
+			return nil, errors.GetAPIError(errors.ErrNoSuchUpload)
+		}
+		return nil, err
+	}
+	if ids == nil {
+		return nil, errors.GetAPIError(errors.ErrNoSuchUpload)
+	}
+	meta, err := n.objectHead(ctx, p.Bkt.CID, ids[0])
+	if err != nil {
+		return nil, err
+	}
+	return objInfoFromMeta(p.Bkt, meta), nil
+
+}
+
+func (n *layer) getUploadParts(ctx context.Context, uploadID, key string, bktInfo *data.BucketInfo) ([]*data.ObjectInfo, error) {
+	f := &findParams{
+		filters: []*filter{{attr: UploadIDAttributeName, val: uploadID}},
+		cid:     bktInfo.CID,
+		prefix:  "",
+	}
+
+	ids, err := n.objectSearch(ctx, f)
+	if err != nil {
+		if errors.IsS3Error(err, errors.ErrNoSuchKey) {
+			return nil, errors.GetAPIError(errors.ErrNoSuchUpload)
+		}
+		return nil, err
+	}
+
+	res := make([]*data.ObjectInfo, len(ids))
+
+	for _, id := range ids {
+		meta, err := n.objectHead(ctx, bktInfo.CID, id)
+		if err != nil {
+			n.log.Warn("couldn't head a part of upload",
+				zap.Stringer("object id", id),
+				zap.Stringer("bucket id", bktInfo.CID),
+				zap.Error(err))
+			continue
+		}
+		attrs := userHeaders(meta.Attributes())
+		if attrs[UploadKeyAttributeName] == key {
+			num, err := strconv.Atoi(attrs[UploadPartNumberAttributeName])
+			if err != nil {
+				n.log.Warn("could not cast part number",
+					zap.Stringer("object id", id),
+					zap.Stringer("bucket id", bktInfo.CID),
+					zap.Error(err))
+				continue
+			}
+			res[num] = objInfoFromMeta(bktInfo, meta)
```

Comment:
We cannot rely on the part number from attribute because we can have only one part with number 17 for example. It will cause panic.